File: oauth2-extended-authorization-server/authorization-server/src/main/java/com/onlyteo/sandbox/service/JwkService.java
package com.onlyteo.sandbox.service;

import com.nimbusds.jose.jwk.JWK;
import com.nimbusds.jose.jwk.JWKSelector;
import com.nimbusds.jose.jwk.source.JWKSource;
import com.nimbusds.jose.proc.SecurityContext;
import com.onlyteo.sandbox.properties.AppSecurityProperties;

import java.util.ArrayList;
import java.util.List;

public class JwkService implements JWKSource<SecurityContext> {

    private final AppSecurityProperties securityProperties;
    private final KeyService keyService;

    public JwkService(final AppSecurityProperties securityProperties,
                      final KeyService keyService) {
        this.securityProperties = securityProperties;
        this.keyService = keyService;
    }

    @Override
    public List<JWK> get(final JWKSelector jwkSelector, final SecurityContext context) {
        var jwkList = new ArrayList<JWK>();
        securityProperties.getKeys().stream()
                .map(key -> keyService.readPrivateRsaKey(key.getKeyId(), key.getPublicKeyLocation(), key.getPrivateKeyLocation()))
                .filter(jwkSelector.getMatcher()::matches)
                .forEach(jwkList::add);
        return jwkList;
    }
}

File: spring-boot-oauth2-authorization-server-extended/authorization-server/src/main/kotlin/com/onlyteo/sandbox/service/JwkService.kt
package com.onlyteo.sandbox.service

import com.nimbusds.jose.jwk.JWK
import com.nimbusds.jose.jwk.JWKSelector
import com.nimbusds.jose.jwk.source.JWKSource
import com.nimbusds.jose.proc.SecurityContext
import com.onlyteo.sandbox.properties.AppSecurityProperties

class JwkService(
    private val securityProperties: AppSecurityProperties,
    private val keyService: KeyService
) : JWKSource<SecurityContext> {

    override fun get(selector: JWKSelector, context: SecurityContext?): MutableList<JWK> {
        val jwkList = ArrayList<JWK>()
        securityProperties.keys.stream()
            .map { key ->
                keyService.readPrivateRsaKey(
                    key.keyId,
                    key.publicKeyLocation,
                    key.privateKeyLocation
                )
            }
            .filter { key ->
                selector.matcher.matches(key)
            }
            .forEach { key -> jwkList.add(key) }
        return jwkList
    }
}